### ISSUE TITLE
CI cluster teardown failed as of: AWS API request failed (attempt 11): Socket timed out without establishing a connection (retry in 177147 seconds)

### ISSUE THREAD

**Issue by sreis** [opstrace/opstrace#511]
* **Console Output:**

CI runs are stuck in a loop trying to list the clusters during the destroy step.

```
 AWS API request failed (attempt 11): Socket timed out without establishing a connection (retry in 177147 seconds)
```
* **Build Links**:

https://buildkite.com/opstrace/scheduled-main-builds/builds/1955#c33a1842-4044-4a19-a170-a53b39fbce5c/6438-6558
https://buildkite.com/opstrace/scheduled-main-builds/builds/1956#ac4f9ee2-8f5c-4cbd-8328-cf7a0e377c65/6497-6617
https://buildkite.com/opstrace/scheduled-main-builds/builds/1957#0aa1ed89-8426-4900-a4a4-ae2108f65d25/6472-6573
https://buildkite.com/opstrace/scheduled-main-builds/builds/1958#dc1d9448-76a8-4204-94f3-e51abf1f33bd/6598-6722

**Comment by jgehrcke** [comment]
Notes in random order:

* Educated speculation: based on past experience, this happens when AWS API endpoints in and around Asia are not TCP-reachable.
* Error messages: good example for the idea that when an HTTP request fails with a specific error, that logging just the error isn't ideal -- the request should be logged to, in the same log message, so that one knows immediate _which_ request failed. Better example, in our own logging implementation around HTTP requests: [here](https://github.com/opstrace/opstrace/issues/424#issue-816452503).
* good example for an "exponential backoff retrying strategy" not being the holy grail: needs proper parameters, and an upper bond -- but this was also a copy/paste error, see https://github.com/opstrace/opstrace/commit/0199607019d5d4a64e2aae5c4f6f90f933a8c9f7#r48544452
* Relevant connect() timeout constant: https://github.com/opstrace/opstrace/blob/f32fa9ab692bf4569dc11410b49b1313ca2baa3e/lib/aws/src/util.ts#L53

Working on a patch towards better logging, and more sane retrying behavior.

**Comment by jgehrcke** [comment]
> Error messages: good example for the idea that when an HTTP request fails with a specific error, that logging just the error isn't ideal -- the request should be logged to, in the same log message, so that one knows immediate which request failed. Better example, in our own logging implementation around HTTP requests: here.

The AWS SDK throws away the request information when constructing the error object here: https://github.com/aws/aws-sdk-js/blob/909cfe3dad38a21e3a4236d7d2cfa86a81681520/lib/util.js#L904

From the error callback, there is no way to retrieve the request URL.

Could submit another patch, of course.

**Comment by jgehrcke** [comment]
An improvement: https://github.com/opstrace/opstrace/pull/512

Of course this does not help when the underlying problem here is actually AWS API endpoints practically being unreachable. In those cases, the design of the "look cluster up across all regions" is a bit stuck between a rock and a hard place -- what if certain regions don't provide a response? Fallback to assuming that no clusters are in those regions is _not_ an obviously good decision.

But we'll cross this bridge when this really becomes important : ).

**Comment by jgehrcke** [comment]
If certain regions are of little value to us and often make problems: we can exclude them from the set of 'supported regions' for now.

**Comment by jgehrcke** [comment]
High level point of view:

The first step of destroy on AWS is to determine the AWS region the cluster is in.

For that, AWS API calls needs to succeed across all AWS regions.

When only one region fails to provide a response, then the region lookup is considered failed, and the corresponding destroy operation failed.

Some AWS regions were making problems here in the recent hours.. speculation: probably some in and around Asia, where I've seen TCP connect() problems before -- weird network routes sometimes.
Action items:

- Found a retrying behavior weakness. #512.
- Do not use the region lookup in destroy in CI. Can use --region. Pro: do not have this instability. Con: code path not exercised. Difficult decision.
- Do not support all AWS regions. The ones that make problems here are often in Asia. Of course we also want users in Asia. And not supporting those has a weird taste -- feels like discrimination.

**Comment by jgehrcke** [comment]
note to self: important question: is the region lookup subject to global timeout control?

**Comment by sreis** [comment]
```
[2021-03-26T13:01:09Z] 2021-03-26T13:01:09.744Z info: AWS API request failed (attempt 14): Socket timed out without establishing a connection (retry in 30.00 seconds)
[2021-03-26T13:01:34Z] 2021-03-26T13:01:34.015Z debug: aws sdk: [AWS eks undefined 395.473s 15 retries] listClusters({})
[2021-03-26T13:01:34Z] 2021-03-26T13:01:34.015Z debug: this does not appear to be an AWS API error
```
https://buildkite.com/opstrace/scheduled-main-builds/builds/1988#ed0f8fd0-2025-46f3-bd9a-9c2786f0c104/6543-6642

**Comment by jgehrcke** [comment]
the retry delay problem has been addressed, and #765 has been closed by now -- closing this one here, too.
